schema/appdata/data.go

```diff
@@ -52,16 +58,23 @@ type EventData struct {
 	// MsgIndex is the index of the message in the transaction to which this event is associated.
 	// If TxIndex is negative, this index could correspond to the index of the message in
 	// begin or end block processing if such indexes exist, or it can be set to zero.
-	MsgIndex uint32
+	MsgIndex int32
 
 	// EventIndex is the index of the event in the message to which this event is associated.
-	EventIndex uint32
+	EventIndex int32
 
 	// Type is the type of the event.
 	Type string
 
-	// Data is the JSON representation of the event data. It should generally be a JSON object.
+	// Data lazily returns the JSON representation of the event.
 	Data ToJSON
+
+	// Attributes lazily returns the key-value attribute representation of the event.
+	Attributes ToEventAttributes
+}
```

Comment:
I would propose we use the above `Event` structure in STF and app manager instead of what we have now. It can support either key-value events or JSON events lazily and tracks tx, msg and event indexes directly (rather than pushed down into string formatted attributes).